common/nexus/incoming_service_registry.go

```diff
@@ -137,67 +134,47 @@ func (r *IncomingServiceRegistry) Get(ctx context.Context, id string) (*nexus.In
 	return service, nil
 }
 
-// initializeServicesWithTimeout initializes service data or waits until the context deadline if another
-// thread is initializing service data. Data initialization happens in the background and will continue
-// even if the original request context times out.
-func (r *IncomingServiceRegistry) initializeServicesWithTimeout(ctx context.Context) error {
-	var initErr error
-	waitCh := make(chan struct{})
-
-	go func() {
-		r.dataLock.Lock()
-		defer r.dataLock.Unlock()
-		defer close(waitCh)
-		initCtx, cancel := context.WithTimeout(context.Background(), r.config.initializeServicesTimeout())
-		defer cancel()
-		initErr = r.initializeServicesLocked(initCtx)
-	}()
-
+func (r *IncomingServiceRegistry) waitUntilInitialized(ctx context.Context) error {
 	select {
-	case <-waitCh:
-		return initErr
+	case <-r.serviceDataReady:
+		return nil
 	case <-ctx.Done():
 		return ctx.Err()
 	}
 }
 
-// initializeServicesLocked initializes the in-memory view of service data.
-// It first tries to load from matching service and falls back to querying persistence directly if matching is unavailable.
-func (r *IncomingServiceRegistry) initializeServicesLocked(ctx context.Context) error {
-	if r.serviceDataReady.Ready() {
-		// Indicates service data was loaded by another thread while waiting for write lock.
-		return nil
-	}
-
-	tableVersion, services, err := r.getAllServicesMatching(ctx)
-	if err != nil {
-		// Fallback to persistence on matching error during initial load.
-		r.logger.Error("error from matching when initializing Nexus incoming service cache", tag.Error(err))
-		tableVersion, services, err = r.getAllServicesPersistence(ctx)
-		if err != nil {
-			return err
-		}
-	}
-
-	r.tableVersion = tableVersion
-	r.services = services
-	r.serviceDataReady.Set(struct{}{}, nil)
-
-	return nil
-}
-
 func (r *IncomingServiceRegistry) refreshServicesLoop(ctx context.Context) error {
-	// Wait for service data to be initialized before starting long poll loop.
-	if _, err := r.serviceDataReady.Get(ctx); err != nil {
-		return err
-	}
-
+	hasLoadedServiceData := false
 	minWaitTime := r.config.refreshMinWait()
 
 	for ctx.Err() == nil {
+		// Wait for Nexus APIs to be enabled.
+		if !r.config.nexusAPIsEnabled() {
+			if hasLoadedServiceData {
+				// Nexus APIs were previously enabled and service data loaded, so make future requests wait for
+				// reload once the FF is re-enabled.
+				r.serviceDataReady = make(chan struct{})
+			}
+			hasLoadedServiceData = false
+			common.InterruptibleSleep(ctx, r.config.refreshMinWait())
```

Comment:
This is probably fine but ideally we would have a way to subscribe to DC changes (@dnr)